### PR TITLE
8314063 : The socket is not closed in Connection::createSocket when the handshake failed for LDAP connection

### PR DESCRIPTION
**Pull request by weibxiao**
Please refer to JDK-8314063.

The failure scenario is due to the setting of connection timeout. It is either too small or not an optimal value for the system. When the client tries to connect to the server with LDAPs protocol. It requires the handshake after the socket is created and connected, but it fails due to connection timeout and leaves the socket open. It is not closed properly due to the exception handling in the JDK code.

The change is adding a try/catch block and closing the socket in the catch block,  and the format of the code got changed consequently.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314063](https://bugs.openjdk.org/browse/JDK-8314063): The socket is not closed in Connection::createSocket when the handshake failed for LDAP connection (**Bug** - P3)

### Reviewers
 * [Aleksei Efimov](https://openjdk.org/census#aefimov) (@AlekseiEfimov - **Reviewer**)
 * [Mark Sheppard](https://openjdk.org/census#msheppar) (@msheppar - **Reviewer**) ⚠️ Review applies to [71bca08c](https://git.openjdk.org/jdk/pull/15294/files/71bca08c8e6c7ebb4e16da8743e2caba333740e0)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15294/head:pull/15294` \
`$ git checkout pull/15294`

Update a local copy of the PR: \
`$ git checkout pull/15294` \
`$ git pull https://git.openjdk.org/jdk.git pull/15294/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15294`

View PR using the GUI difftool: \
`$ git pr show -t 15294`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15294.diff">https://git.openjdk.org/jdk/pull/15294.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15294#issuecomment-1679343291)

### CODE REVIEW

**Review by AlekseiEfimov — COMMENTED** [review]
Hello Weibing,
The approach taken here seems resonable to me. Please find my comments on the fix and the test below:

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        }
+
+        public boolean isExceptionThrown() {
+            return exceptionThrown;
+        }
+
+        @Override
+        public void run() {
+            try (Socket socket = serverSocket.accept()) {
+                Thread.sleep(10000);
```
Comment (msheppar) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:173]: What's the purpose of the sleep ? 
Regardless, based on the test semantics alluded in the test name, the server should never enter the read block. So is this code redundant? Or is it there just in case the accept and the handshake succeeds?

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 2]
```diff
@@ -278,93 +278,100 @@ private InetSocketAddress createInetSocketAddress(String host, int port) {
...
+                    sslSocket.setSoTimeout(socketTimeout);
+                }
             }
-            setHandshakeCompletedListener(sslSocket);
-            if (connectTimeout > 0) {
-                int socketTimeout = sslSocket.getSoTimeout();
-                sslSocket.setSoTimeout(connectTimeout); // reuse full timeout value
-                sslSocket.startHandshake();
-                sslSocket.setSoTimeout(socketTimeout);
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (msheppar) [reply src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:369]: But the catch is on generalized Exception, and SecutityException is a subclass, so it is covered, n'est-ce pas?

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 3]
```diff
@@ -278,93 +278,100 @@ private InetSocketAddress createInetSocketAddress(String host, int port) {
...
+            // 8314063 the socket is not closed after the failure of handshake
+            if (socket != null && !socket.isClosed()) {
+                socket.close();
```
[the remark this answers is not in the document]
Comment (msheppar) [reply src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:372]: a closeConnectionSocket (IIRC) was introduced in previous change which handled any IOExceptions on the close. This could be used here.
But some how that seems to have been renamed !!

709     private void closeOpenedSocket() {
 710         try {
 711             sock.close();
 712         } catch (IOException ioEx) {
 713             if (debug) {
 714                 System.err.println("Connection.closeConnectionSocket: Socket close problem: " + ioEx);
 715                 System.err.println("Socket isClosed: " + sock.isClosed());
 716             }
 717         }
 718     }

**Comment by weibxiao** [comment]
1) Refactor Connection::createSocket
2) Update the test code.

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java (hunk 1 above)
Comment (msheppar) [reply test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:177]: OK, So the SSL handshake must take place within the getInputStream call rather than the accept, as such you are triggering a timeout on the SSL handshake.

**Comment by msheppar** [comment]
together with a refactor extact methods on the createSocket method, the existing logic could be simplified something like as follows:


            InetSocketAddress endpoint =
               createInetSocketAddress(ldapServerHost, ldapServerPort);

            if (socketFactory != null) {
                 // further refine with refactor extract method createConnectionSocket(InetSocketAddress serverEndpoint, String factoryClassName, int timeout)

                // create the factory

                @SuppressWarnings("unchecked")
                Class<? extends SocketFactory> socketFactoryClass =
                        (Class<? extends SocketFactory>) Obj.helper.loadClass(socketFactory);
                Method getDefault =
                        socketFactoryClass.getMethod("getDefault", new Class<?>[]{});
                SocketFactory factory = (SocketFactory) getDefault.invoke(null, new Object[]{});
 
                // create unconnected socket
                socket = factory.createSocket();

               if (connectTimeout > 0) {                 
                  if (debug) {
                        System.err.println("Connection.createSocket: creating socket with " +
                                "a timeout using supplied socket factory");
                    }

                    // connect socket
                    socket.connect(endpoint, connectTimeout);
               } else {
                   if (debug) {
                        System.err.println("Connection.createSocket: creating socket using " +
                                "supplied socket factory");
                    }
                    // connect socket
                    socket.connect(endpoint);
               }
             } else { // NO SocketFactory
                // further refine with refactor extract method createConnectionSocket(InetSocketAddress serverEndpoint,  int timeout)
                socket = new Socket();
                if (connectTimeout > 0) {
                    if (debug) {
                        System.err.println("Connection.createSocket: creating socket with " +
                                "a timeout");
                    }
                    socket.connect(endpoint, connectTimeout);
                } else {
                   if (debug) {
                        System.err.println("Connection.createSocket: creating socket");
                   }
                   socket.connect(endpoint);
              }

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 4]
```diff
@@ -278,79 +278,47 @@ private InetSocketAddress createInetSocketAddress(String host, int port) {
...
-
-                    if (debug) {
-                        System.err.println("Connection: creating socket with " +
-                            "a timeout");
-                    }
-                    socket.connect(endpoint, connectTimeout);
+        try {
+            // create the socket with factory
+            if (socketFactory != null) {
+                socket = createSocketWithFactory (host, port, socketFactory, connectTimeout) ;
```
Comment (msheppar) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:287]: you could use method overloading :-)

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 5]
```diff
@@ -366,6 +334,68 @@ private Socket createSocket(String host, int port, String socketFactory,
                 sslSocket.setSoTimeout(socketTimeout);
             }
         }
+    }
+
+    // create the socket without the factory
+    private Socket createSocketWithoutFactory (String host, int port, int connectTimeout) throws Exception {
```
Comment (msheppar) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:340]: i think the logic can be streamlined as per comment in main converstaion

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 6]
```diff
@@ -366,6 +334,68 @@ private Socket createSocket(String host, int port, String socketFactory,
...
+                socketFactoryClass.getMethod("getDefault", new Class<?>[]{});
+        SocketFactory factory = (SocketFactory) getDefault.invoke(null, new Object[]{});
+
+        // create the socket
+        if (connectTimeout > 0) {
+
+            InetSocketAddress endpoint =
+                    createInetSocketAddress(host, port);
+            // unconnected socket
+            socket = factory.createSocket();
```
Comment (msheppar) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:381]: i think the logic can be streamlined as per comment in main conversation

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        }
+
+        @Override
+        public void close() throws IOException {
+            closeMethodCalled++;
+            super.close();
+        }
+    }
+
+    private static void setKeyStore() {
```
Comment (vyommani) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:143]: Overall code changes looks OK to me,  did you consider creating the self signed certificates programmatically and use them in  your test ?. Please  look into the existing javax.net.ssl tests how to do this.

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlekseiEfimov) [reply src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java]: Thanks Mark. Correct - SecurityException is a subclass of Exception, scratch my previous comment. We just need to double check `catch Exception` covers all thrown exception types in this code block.

**Comment by weibxiao** [comment]
1) Refactor the code according to the suggestion
2) Removed the pre-created key store and re-create it at the runtime.

**Comment by weibxiao** [comment]
1) Please ignore the previous change for Connection.java
2) Refactor the method of createSocket
3) Re-write closedOpenedSocket, and remove closeSocket method.

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 8]
```diff
@@ -278,79 +278,86 @@ private InetSocketAddress createInetSocketAddress(String host, int port) {
...
+        try {
+            InetSocketAddress endpoint =
+                    createInetSocketAddress(host, port);
+
+            if (socketFactory != null) {
+                // create a connected socket with factory
+                socket = createConnectionSocket(endpoint, socketFactory, connectTimeout);
+            } else { // NO SocketFactory
+                // create a connected socket without factory
+                createConnectionSocket(endpoint, connectTimeout);
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:293]: A socket returned by `createConnectionSocket` here needs to be saved to the `socket` variable - currently almost all tests that are not using a custom socket factory - fails with:
```
Caused by: java.lang.NullPointerException: Cannot invoke "java.net.Socket.getInputStream()" because "this.sock" is null
	at java.naming/com.sun.jndi.ldap.Connection.<init>(Connection.java:238)
```

**Review by AlekseiEfimov — CHANGES_REQUESTED** [review]

**Comment by weibxiao** [comment]
1) Refactor the code again, and simplify the code to handle the setting of connection time
2) Add more test cases
3) Bring back the pre-created keystore. I could not find a good example to create the key store at runtime.

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    private static String SOCKET_CLOSED_MSG = "The socket has been closed.";
+    private static String SOCKET_NOT_CLOSED_MSG = "The socket was not closed.";
+
+    public static void main(String args[]) throws Exception {
+
+        // Set the keystores
+        setKeyStore();
+        // start the test server first.
+        boolean serverSlowDown = false;
+        if(args.length ==2 ) {
```
Comment (turbanoff) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:77]:
```suggestion
        if (args.length ==2 ) {
```

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+            serverSlowDown = Boolean.valueOf(args[1]);
+        } else {
+            if(args.length ==1 ) {
```
Comment (turbanoff) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:80]:
```suggestion
            if (args.length ==1) {
```

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+                .loopback()
+                .port(server.getPortNumber())
+                .buildUnchecked().toString());
+        if (args.length == 2 &&
+                args[0].contains("LdapSSLHandshakeFailureTest")) {
+            env.put("java.naming.ldap.factory.socket", args[0]);
+        }
+        env.put("java.naming.ldap.version", "3");
+        if (args.length == 2 ) {
+            if( args[0].contains("LdapSSLHandshakeFailureTest")) {
```
Comment (turbanoff) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:99]:
```suggestion
            if (args[0].contains("LdapSSLHandshakeFailureTest")) {
```

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        } catch (Exception e) {
+            if (CustomSocketFactory.customSocket.closeMethodCalledCount() > 0
+                    && args[0].equals("LdapSSLHandshakeFailureTest$CustomSocketFactory")
+                    && Boolean.valueOf(args[1])) {
+                System.out.println(SOCKET_CLOSED_MSG);
+            } else {
+                throw e;
+            }
+        } finally {
+            if(ctx != null)
```
Comment (turbanoff) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:122]:
```suggestion
            if (ctx != null)
```

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+                .scheme("ldaps")
+                .loopback()
+                .port(server.getPortNumber())
+                .buildUnchecked().toString());
+        if (args.length == 2 &&
+                args[0].contains("LdapSSLHandshakeFailureTest")) {
+            env.put("java.naming.ldap.factory.socket", args[0]);
+        }
+        env.put("java.naming.ldap.version", "3");
+        if (args.length == 2 ) {
```
Comment (turbanoff) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:98]:
```suggestion
        if (args.length == 2) {
```

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+        server.start();
+        Hashtable<String, Object> env = new Hashtable<>();
+        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
+        env.put(Context.PROVIDER_URL, URIBuilder.newBuilder()
+                .scheme("ldaps")
+                .loopback()
+                .port(server.getPortNumber())
+                .buildUnchecked().toString());
+        if (args.length == 2 &&
+                args[0].contains("LdapSSLHandshakeFailureTest")) {
```
Comment (AlekseiEfimov) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:94]: This `arg[0]` check is done twice in the test - it can be save to a boolean variable:
```
boolean hasCustomSocketFactory = args[0]
        .equals("LdapSSLHandshakeFailureTest$CustomSocketFactory");
```
and then used to the checks

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        // start the test server first.
+        boolean serverSlowDown = false;
+        if(args.length ==2 ) {
+            serverSlowDown = Boolean.valueOf(args[1]);
+        } else {
+            if(args.length ==1 ) {
+                serverSlowDown = Boolean.valueOf(args[0]);
+            }
+        }
+        TestServer server = new TestServer( serverSlowDown );
```
Comment (AlekseiEfimov) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:84]: The server instance can be used in try-with-resouces:
```
        try (TestServer server = new TestServer(serverSlowDown)) {
            server.start();
            env.put(Context.PROVIDER_URL, URIBuilder.newBuilder()
                    .scheme("ldaps")
                    .loopback()
                    .port(server.getPortNumber())
                    .buildUnchecked().toString());
```

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+ * @run main/othervm LdapSSLHandshakeFailureTest 2000 false
+ * @run main/othervm LdapSSLHandshakeFailureTest 0 true
+ * @run main/othervm LdapSSLHandshakeFailureTest 0 false
+ * @run main/othervm LdapSSLHandshakeFailureTest true
+ * @run main/othervm LdapSSLHandshakeFailureTest false
+ */
+
+public class LdapSSLHandshakeFailureTest {
+    private static String SOCKET_CLOSED_MSG = "The socket has been closed.";
+    private static String SOCKET_NOT_CLOSED_MSG = "The socket was not closed.";
```
Comment (AlekseiEfimov) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:69]: `SOCKET_NOT_CLOSED_MSG` - not used and can be removed

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+        }
+
+        @Override
+        public void close() throws java.io.IOException {
+            closeMethodCalled++;
+            super.close();
+        }
+    }
+
+    private static void setKeyStore() throws Exception {
```
Comment (AlekseiEfimov) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:178]: `Exception` is not thrown by this method and can be removed

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+            setDaemon(true);
+        }
+
+        public int getPortNumber() {
+            return PORT;
+        }
+
+        @Override
+        public void run() {
+            try (Socket socket = serverSocket.accept(); InputStream in = socket.getInputStream();
```
Comment (AlekseiEfimov) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:211]: formatting: this can be made shorter by putting `InputStream in = socket.getInputStream();`  on a separate line

File: test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+                 OutputStream out = socket.getOutputStream()) {
+                if (isForceToSleep) {
+                    Thread.sleep(5000);
+                }
+                byte[] bindResponse = {0x30, 0x0C, 0x02, 0x01, 0x01, 0x61, 0x07, 0x0A, 0x01, 0x00, 0x04, 0x00, 0x04, 0x00};
```
Comment (AlekseiEfimov) [inline test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java:216]: formatting: can be splitted in two lines to make it shorter

**Comment by weibxiao** [comment]
1) Reformat the test code
2) Add one more comment in Connection.java
3) Updated the test code base on the review comment

**Review by AlekseiEfimov — COMMENTED** [review]
The latest version looks fine to me:
New `createConnectionSocket` methods make code more readable while 
maintaining the same codeflow in respect of connection timeout and
socket factory method calls.
Existing LDAP tests show no problems with the proposed fix.
 
Few minor formatting suggestions:
 
Remove unused imports in the test:
```
 import java.io.File;
-import java.io.FileInputStream;
-import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
@@ -36,7 +34,6 @@ import javax.naming.ldap.InitialLdapContext;
 import javax.naming.ldap.LdapContext;
 import javax.naming.Context;
 import java.net.SocketException;
-import java.security.KeyStore;
```

Remove extra space:
```
-        public CustomSocket () {
+        public CustomSocket() {
```

Remove redundant default value:
```
-        private boolean isForceToSleep = false;
+        private boolean isForceToSleep;
```

**Comment by weibxiao** [comment]
Formatted the code according to the comments.

**Comment by msheppar** [comment]
Resuggesting restructure to createConnectionSocket methods -- to instantiate Socket explicitly in one place and to explicitly use connect in all cases

`
   private Socket createConnectionSocket (String host, int port, int connectTimeout) throws Exception {
        InetSocketAddress endpoint = createInetSocketAddress(host, port);
        Socket socket = new Socket();

        if (connectTimeout > 0) {
            if (debug) {
                System.err.println("Connection: creating socket with " +
                        "a timeout");
            }
            socket.connect(endpoint, connectTimeout);
        } else {
            if (debug) {
                System.err.println("Connection: creating socket");
            }
            // connected socket
            socket.connect(endpoint);
        }
        return socket;
    }


    // create the socket with the provided factory
    private Socket createConnectionSocket(String host, int port, String socketFactory,
                                           int connectTimeout) throws Exception {
        @SuppressWarnings("unchecked")
        Class<? extends SocketFactory> socketFactoryClass = (Class<? extends SocketFactory>)
                Obj.helper.loadClass(socketFactory);
        Method getDefault =
                socketFactoryClass.getMethod("getDefault", new Class<?>[]{});
        SocketFactory factory = (SocketFactory) getDefault.invoke(null, new Object[]{});
        InetSocketAddress endpoint =
                  createInetSocketAddress(host, port);
        // create unconnected socket
        Socket socket = factory.createSocket();

        // create the socket
        if (connectTimeout > 0) {
            if (debug) {
                System.err.println("Connection: creating socket with " +
                        "a timeout using supplied socket factory");
            }
            // connected socket
            socket.connect(endpoint, connectTimeout);
        } else {
            if (debug) {
                System.err.println("Connection: creating socket using " +
                        "supplied socket factory");
            }
            // connected socket
            socket.connect(endpoint);
        }
        return socket;
    }

`

**Comment by weibxiao** [comment]
If the custom factory does not implement the method creatSocket (no method parameters), it will fail for creating the socket. The SocketFactory::createSocket is not defined as an abstract method, instead, it throws UnsupportedOperationException. The subclass is not required to implement it.

**Comment by msheppar** [comment]
Such a restriction appears to be a flaw in the design of SocketFactory and its use in jndi ldap.
Consider the existing updated code  and the inference of your assertion:
 ```
       // create the socket
        if (connectTimeout > 0) {
            InetSocketAddress endpoint =
                    createInetSocketAddress(host, port);
            // unconnected socket
            socket = factory.createSocket();
            // connected socket
            socket.connect(endpoint, connectTimeout);
            if (debug) {
                System.err.println("Connection: creating socket with " +
                        "a timeout using supplied socket factory");
            }
        } else {
            // continue (but ignore connectTimeout)
            if (socket == null) {
                // connected socket
                socket = factory.createSocket(host, port);
                if (debug) {
                    System.err.println("Connection: creating socket using " +
                            "supplied socket factory");
                }
            }
        }``
```
Your assertion is, "If the custom factory does not implement the method creatSocket (no method parameters), it will fail for creating the socket."
Thus, the first block of the, if (connectTimeout > 0) condition, is problematic. This assumes that if the timeout is > 0, then the factory will implement the no args factory method createSocket(). BUT that means that an implementor  of a customer factory would need to have knowledge of the internal implemntation of this class, and ensure no args factory method createSocket is implemented. If an implementor has knowledge to implement the no args factory method when timeout > 0, then they can also have knowldge to implement the no args factory method when the timeout is 0 or < 0, as suggested by the restructure.

Additionally, this would not prohibit a restructure of the non factory createConnectionSocket.

**Comment by AlekseiEfimov** [comment]
May I suggest a slightly different approach for refactoring of the Connection class code here:
Instead of having two methods for creating a `Socket` with OR without socket factory the method to select a socket factory can be introduced: 
If a socket factory class name is not set - the factory returned by `SocketFactory.getDefault()` will be used (which is identical to `s = new Socket()` for calls with timeout set; and `s = new Socket(host, port)` for calls with no timeout set):

```    private SocketFactory getSocketFactory(String socketFactoryName) throws Exception {
    if (socketFactoryName == null) {
        if (debug) {
            System.err.println("Connection: using default SocketFactory");
        }
        return SocketFactory.getDefault();
    } else {
        if (debug) {
            System.err.println("Connection: loading supplied SocketFactory: " + socketFactoryName);
        }
        @SuppressWarnings("unchecked")
        Class<? extends SocketFactory> socketFactoryClass =
                (Class<? extends SocketFactory>) Obj.helper.loadClass(socketFactoryName);
        Method getDefault =
                socketFactoryClass.getMethod("getDefault");
        SocketFactory factory = (SocketFactory) getDefault.invoke(null, new Object[]{});
        return factory;
    }
}
```

It then can be used to create a connection socket:
```
    private Socket createConnectionSocket(String host, int port, SocketFactory factory, int connectTimeout) throws Exception {
        Socket socket = null;

        if (connectTimeout > 0) {
            // create unconnected socket and then connect it if timeout
            // is supplied
            InetSocketAddress endpoint =
                    createInetSocketAddress(host, port);
            // unconnected socket
            socket = factory.createSocket();
            // connect socket with a timeout
            socket.connect(endpoint, connectTimeout);
            if (debug) {
                System.err.println("Connection: creating socket with " +
                        "a connect timeout");
            }
        }
        if (socket == null) {
            // create connected socket
            socket = factory.createSocket(host, port);
            if (debug) {
                System.err.println("Connection: creating connected socket with" +
                        " no connect timeout");
            }
        }
        return socket;
    }
```

And then `createSocket` can be changed to:
```
    private Socket createSocket(String host, int port, String socketFactory,
                                int connectTimeout) throws Exception {

        SocketFactory factory = getSocketFactory(socketFactory);
        assert factory != null;
        Socket socket = createConnectionSocket(host, port, factory, connectTimeout);

        // the handshake for SSL connection with server and reset timeout for the socket
        try {
            initialSSLHandshake(socket, connectTimeout);
        } catch (Exception e) {
            // 8314063 the socket is not closed after the failure of handshake
            // close the socket while the error happened
            closeOpenedSocket(socket);
            throw e;
        }
        return socket;
    }
```

Also, the new try catch block can be put around initialSSLHandshake method only - it is the only case when socket can be set to a non-null value.

**Comment by msheppar** [comment]
looks like a good approach 👍

**Comment by weibxiao** [comment]
Please see the updated code. Thanks.

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 20]
```diff
@@ -25,98 +25,95 @@
...
 import java.util.concurrent.ExecutionException;
-import javax.net.SocketFactory;
-import javax.net.ssl.SSLParameters;
-import javax.net.ssl.HandshakeCompletedEvent;
-import javax.net.ssl.HandshakeCompletedListener;
-import javax.net.ssl.SSLPeerUnverifiedException;
-import javax.security.sasl.SaslException;
 
 /**
-  * A thread that creates a connection to an LDAP server.
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:62]: The formatting changes here is not related to the code changes under review in this PR. Could you please revert them?

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 21]
```diff
@@ -128,9 +125,9 @@ public final class Connection implements Runnable {
     private boolean v3 = true;       // Set in setV3()
 
     public final String host;  // used by LdapClient for generating exception messages
-                         // used by StartTlsResponse when creating an SSL socket
+    // used by StartTlsResponse when creating an SSL socket
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:128]: Same - reformatting of unrelated code

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 22]
```diff
@@ -128,9 +125,9 @@ public final class Connection implements Runnable {
...
     public final int port;     // used by LdapClient for generating exception messages
-                         // used by StartTlsResponse when creating an SSL socket
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:133]: Same - reformatting of unrelated code

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 23]
```diff
@@ -243,15 +241,15 @@ void setBound() {
             // realException.printStackTrace();
 
             CommunicationException ce =
-                new CommunicationException(host + ":" + port);
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:246]: Same - reformatting of unrelated code

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 24]
```diff
@@ -243,15 +241,15 @@ void setBound() {
...
+                    new CommunicationException(host + ":" + port);
             ce.setRootCause(realException);
             throw ce;
         } catch (Exception e) {
             // We need to have a catch all here and
             // ignore generic exceptions.
             // Also catches all IO errors generated by socket creation.
             CommunicationException ce =
-                new CommunicationException(host + ":" + port);
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:254]: Same - reformatting of unrelated code

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 25]
```diff
@@ -278,79 +276,81 @@ private InetSocketAddress createInetSocketAddress(String host, int port) {
...
+    // the SSL handshake following socket connection as part of the timeout.
+    // So explicitly set a socket read timeout, trigger the SSL handshake,
+    // then reset the timeout.
+    private void initialSSLHandshake(Socket socket, int connectTimeout) throws Exception {
 
-        // For LDAP connect timeouts on LDAP over SSL connections must treat
-        // the SSL handshake following socket connection as part of the timeout.
-        // So explicitly set a socket read timeout, trigger the SSL handshake,
-        // then reset the timeout.
         if (socket instanceof SSLSocket) {
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:354]: instanceof pattern matching can be used here:
```
        if (socket instanceof SSLSocket sslSocket) {
```

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 26]
```diff
@@ -384,15 +383,15 @@ LdapRequest writeRequest(BerEncoder ber, int msgId) throws IOException {
     }
 
     LdapRequest writeRequest(BerEncoder ber, int msgId,
-        boolean pauseAfterReceipt) throws IOException {
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:387]: Reformatting of unrelated code

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 27]
```diff
@@ -384,15 +383,15 @@ LdapRequest writeRequest(BerEncoder ber, int msgId) throws IOException {
...
+                             boolean pauseAfterReceipt) throws IOException {
         return writeRequest(ber, msgId, pauseAfterReceipt, -1);
     }
 
     LdapRequest writeRequest(BerEncoder ber, int msgId,
-        boolean pauseAfterReceipt, int replyQueueCapacity) throws IOException {
+                             boolean pauseAfterReceipt, int replyQueueCapacity) throws IOException {
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:391]: Reformatting of unrelated code

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 28]
```diff
@@ -384,15 +383,15 @@ LdapRequest writeRequest(BerEncoder ber, int msgId) throws IOException {
...
 
         LdapRequest req =
-            new LdapRequest(msgId, pauseAfterReceipt, replyQueueCapacity);
+                new LdapRequest(msgId, pauseAfterReceipt, replyQueueCapacity);
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:394]: Reformatting of unrelated code

File: src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java [hunk 29]
```diff
@@ -616,14 +615,14 @@ private void ldapUnbind(Control[] reqCtls) {
     }
 
     /**
-     * @param reqCtls Possibly null request controls that accompanies the
```
Comment (AlekseiEfimov) [inline src/java.naming/share/classes/com/sun/jndi/ldap/Connection.java:619]: Reformatting of unrelated code

**Review by AlekseiEfimov — COMMENTED** [review]
The last change reformatted several lines unrelated to this change - I've commented on a few occurrences. Could you please keep only changes related to the scope of this PR?

**Comment by weibxiao** [comment]
1) Add server sleeping time through the command line
2) Move out instanceof pattern out of the defined method
3)Extra testings were done.

**Review by msheppar — APPROVED** [review]
